src/backend_dev_setup.md

```diff
@@ -0,0 +1,215 @@
+to your host language. Finally we provide a suggested checklist for your backend.
+It is not automatically generated so when in doubt look at diplomat's [HIR](https://docs.rs/diplomat_core/latest/diplomat_core/hir/index.html)
+
+- [ ] **project structure**: You will need need to test your generated code so you should
+first set up a host language project. It should have all dependencies to be able to interface 
+with native code (or WASM).
+
+## Setting up Basic Code Generation in a Test
+
+Your backend should iterate over all [TypeDefs](https://docs.rs/diplomat_core/latest/diplomat_core/hir/enum.TypeDef.html)
+and generate the required code for these. A good starting point is to create a test for generating a simple opaque struct
+without any methods. Your backend should go in the tool create: create a module `tool/src/{backend}/mod.rs` 
+(make sure you add a line `pub mod backend;` to `tool/src/lib.rs`). Add the following to it
+
+```rs
+use diplomat_core::hir::{OpaqueDef, TypeContext, TypeId};
+
+fn gen_opaque_def(ctx: &TypeContext, type_id: TypeId, opaque_path: &OpaqueDef) -> String {
+    "We'll get to it".into()
+}
+
+#[cfg(test)]
+mod test {
+    use diplomat_core::{
+        ast::{self},
+        hir::{self, TypeDef},
+    };
+    use quote::quote;
+
+    #[test]
+    fn test_opaque_gen() {
+        let tokens = quote! {
+            #[diplomat::bridge]
+            mod ffi {
+
+                #[diplomat::opaque]
+                struct OpaqueStruct;
+
+            }
+        };
+        let item = syn::parse2::<syn::File>(tokens).expect("failed to parse item ");
+
+        let diplomat_file = ast::File::from(&item);
+        let env = diplomat_file.all_types();
+        let attr_validator = hir::BasicAttributeValidator::new("my_backend_test");
+
+        let context = match hir::TypeContext::from_ast(&env, attr_validator) {
+            Ok(context) => context,
+            Err(e) => {
+                for (_cx, err) in e {
+                    eprintln!("Lowering error: {}", err);
+                }
+                panic!("Failed to create context")
+            }
+        };
+
+        let (type_id, opaque_def) = match context
+            .all_types()
+            .next()
+            .expect("Failed to generate first opaque def")
+        {
+            (type_id, TypeDef::Opaque(opaque_def)) => (type_id, opaque_def),
+            _ => panic!("Failed to find opaque type from AST"),
+        };
+
+        let generated = super::gen_opaque_def(&context, type_id, opaque_def);
+
+        insta::assert_snapshot!(generated)
+    }
+}
+```
+
+You can now run 
+```sh
+cargo test -p diplomat-tool -- backend::test --nocapture
+```
+You should also have a generated snapshot `diplomat_tool__backend__test__opaque_gen.snap.new` 
+which you can use to pick up your generated code.
+
+## How to Generate the Dylib
+Now to actually test native methods you will need to create a dynamically linked library 
```

Comment:
issue: not always, no. say that you will need to compile _some_ kind of library: a staticlib, dylib, or wasm lib. And then say here we'll make a dylib.